File: src/two_means/test/test_gmeans.c
#include "gmeans.h"
#include "utils.h"


void test_first_prcomp(double *X, int *pp, int *ck, int *pnk, double *mu, double *vec1, double *vec2) {
  int i, j;
  int p = *pp, nk = *pnk;
  double lambda;

  double **cov = (double **)malloc(sizeof(double *) * p);
  for (i = 0; i < p; i++)
    cov[i] = (double *)malloc(sizeof(double) * p);

  covariance(X, p, ck, nk, mu, cov);

  printf("covariance:\n");
  for (i = 0; i < p; i++) {
    for (j = 0; j < p; j++)
      printf("%2.4f ", cov[i][j]);
    printf("\n");
  }

  lambda = first_prcomp(X, p, ck, nk, mu, cov, vec1, vec2);

  printf("lambda = %.5f\n", lambda);

  for (i = 0; i < p; i++) free(cov[i]);
  free(cov);
}

void test_normalCDF(double *value) {
  printf("cfd = %.5f\n", normalCDF(*value));
}

void test_ADstatic(double *x, int *pn, double *d) {
  int i, n = *pn;
  double *z;

  qsort(x, n, sizeof(double), asc_double);

  z = (double *)malloc(sizeof(double) * n);
  for (i = 0; i < n; i++) z[i] = normalCDF(x[i]);

  *d = ADstatic(z, n);

  free(z);
}

void test_projection(double *X, int *p, int *ck, int *nk, double *vec1, double *vec2, double *prj_x) {
  projection(X, *p, ck, *nk, vec1, vec2, prj_x);
}

File: src/two_means/test/test_gmeans.R
## create share object using the following code in your bash
##  R CMD SHLIB -c test_gmeans.c gmeans.c twomeans.c list.c utils.c

dyn.load("test_gmeans.so")

X <- as.matrix(iris[,-5])
mu <- colMeans(X)
n <- nrow(X)
p <- ncol(X)
ck <- 1:n

ret <- .C("test_first_prcomp",
          X = as.double(t(X)),
          p = as.integer(p),
          ck = as.integer(ck-1),
          nk = as.integer(length(ck)),
          mu = as.double(mu),
          vec1 = double(p),
          vec2 = double(p))

X.cov <- (t(X) - mu) %*% t(t(X) - mu) / n
cov.e <- eigen(X.cov)
cov.e$values[1]
vec - ret$vec1

## normal CFD
ret <- .C("test_normalCFD", value=as.double(1.2))
pnorm(1.2)

## test ADstatic
x <- rnorm(100)
n <- length(x)

ret <- .C("test_ADstatic",
          x=as.double(x),
          n=as.integer(n),
          d=double(1))

z <- pnorm(sort(x))
A <- -n - mean((2 * (1:n) - 1) * (log(z) + log(1-rev(z))))
ret$d - A * (1 + 4/n - 25 / (n*n))

## test projection
X <- as.matrix(iris[, -5])
n <- nrow(X)
p <- ncol(X)
km <- kmeans(X, 2)
vec1 <- km$centers[1,]
vec2 <- km$centers[2,]
ck <- 1:n

ret <- .C("test_projection",
          X = as.double(t(X)),
          p = as.integer(p),
          ck = as.integer(ck-1),
          nk = as.integer(n),
          vec1 = as.double(vec1),
          vec2 = as.double(vec2),
          prj_x = double(n)
          )

vec <- vec1 - vec2
prj_x <- scale(X %*% vec / sum(vec**2))
norm(ret$prj_x - prj_x, "2")

## test gmeans
Q <- MixSim::MixSim(0.05, p=2, K=3, sph=T, hom=F)
A <- MixSim::simdataset(1000, Pi=Q$Pi, Mu=Q$Mu, S=Q$S)
X <- A$X

km <- gmeans(X, debug=1)
